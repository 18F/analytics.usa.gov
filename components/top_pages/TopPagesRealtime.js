import React, { useRef, useEffect } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import renderBlock from "../../js/lib/renderblock";
import { titleExceptions } from "../../js/lib/exceptions";
import barChart from "../../js/lib/barchart";
import formatters from "../../js/lib/formatters";

function TopPagesRealtime({ dataHrefBase, reportFileName }) {
  const dataURL = `${dataHrefBase}/${reportFileName}`;
  const ref = useRef(null);

  useEffect(() => {
    const initRealtimeChart = async () => {
      const result = await d3
        .select(ref.current)
        .datum({
          source: dataURL,
          block: ref.current,
        })
        .call(
          renderBlock
            .loadAndRender()
            .transform((d) => d.data)
            .on("render", (selection) => {
              selection
                .selectAll(".label")
                .each(function (d) {
                  d.text = this.innerText;
                })
                .html("")
                .text((d) => titleExceptions[d.page] || d.page_title);
            })
            .render(
              barChart()
                .label((d) => d.page_title)
                .value((d) => +d.active_visitors)
                .scale((values) =>
                  d3.scale
                    .linear()
                    .domain([0, 1, d3.max(values)])
                    .rangeRound([0, 1, 100]),
                )
                .format(formatters.addCommas),
            ),
        );
      return result;
    };
    initRealtimeChart().catch(console.error);
  }, []);

  return (
    <figure
      className="top-pages__realtime"
      data-source={dataURL}
      data-refresh="15"
      ref={ref}
    >
      <div className="data bar-chart"></div>
    </figure>
  );
}

TopPagesRealtime.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
  reportFileName: PropTypes.string.isRequired,
};

export default TopPagesRealtime;
