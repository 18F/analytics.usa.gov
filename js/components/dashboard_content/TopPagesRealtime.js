import React, { useRef, useEffect, useState } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import DataLoader from "../../lib/data_loader";
import ChartBuilder from "../../lib/chart_helpers/chart_builder";
import { titleExceptions } from "../../lib/exceptions";
import barChart from "../../lib/chart_helpers/barchart";
import formatters from "../../lib/chart_helpers/formatters";

/**
 * Retrieves the realitme top page report from the passed data URL and creates a
 * visualization for the count of users visiting sites for the current agency.
 *
 * @param {object} props the properties for the component
 * @param {string} props.dataHrefBase the URL of the base location of the data
 * to be downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 * @param {string} props.reportFileName the file name of the report to use as a
 * data source for the data visualization.
 * @param {number} props.numberOfListingsToDisplay the count of top page
 * listings to display in the bar chart.
 * @param {number} props.refreshSeconds the number of seconds to wait before
 * refreshing chart data.
 * @returns {import('react').ReactElement} The rendered element
 */
function TopPagesRealtime({
  dataHrefBase,
  reportFileName,
  numberOfListingsToDisplay,
  refreshSeconds,
}) {
  const dataURL = `${dataHrefBase}/${reportFileName}`;
  const ref = useRef(null);
  const [topPagesData, setTopPagesData] = useState(null);

  useEffect(() => {
    const initTopPagessChart = async () => {
      if (!topPagesData) {
        const data = await DataLoader.loadJSON(dataURL);
        await setTopPagesData(data);
        // Refresh data every interval. useEffect will run and update the chart
        // when the state is changed.
        setInterval(() => {
          DataLoader.loadJSON(dataURL).then((data) => {
            setTopPagesData(data);
          });
        }, refreshSeconds * 1000);
      } else {
        const chartBuilder = new ChartBuilder();
        await chartBuilder
          .setElement(ref.current)
          .setData(topPagesData)
          .setTransformer((d) => d.data.slice(0, numberOfListingsToDisplay))
          .setRenderer((selection) => {
            const barchartRenderer = barChart()
              .label((d) => d.page_title)
              .value((d) => +d.active_visitors)
              .scale((values) =>
                d3.scale
                  .linear()
                  .domain([0, 1, d3.max(values)])
                  .rangeRound([0, 1, 100]),
              )
              .format(formatters.addCommas);
            barchartRenderer(selection);

            selection
              .selectAll(".label")
              .each(function (d) {
                d.text = this.innerText;
              })
              .html("")
              .text((d) => titleExceptions[d.page] || d.page_title);
          })
          .build();
      }
    };
    initTopPagessChart().catch(console.error);
  }, [topPagesData]);

  return (
    <figure className="top-pages__realtime-bar-chart" ref={ref}>
      <div className="data bar-chart"></div>
    </figure>
  );
}

TopPagesRealtime.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
  reportFileName: PropTypes.string.isRequired,
  numberOfListingsToDisplay: PropTypes.number.isRequired,
  refreshSeconds: PropTypes.number.isRequired,
};

export default TopPagesRealtime;
