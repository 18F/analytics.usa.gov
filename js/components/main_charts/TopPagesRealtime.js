import React, { useRef, useEffect } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import renderBlock from "../../lib/chart_helpers/renderblock";
import { titleExceptions } from "../../lib/exceptions";
import barChart from "../../lib/chart_helpers/barchart";
import formatters from "../../lib/chart_helpers/formatters";

/**
 * Retrieves the realitme top page report from the passed data URL and creates a
 * visualization for the count of users visiting sites for the current agency.
 *
 * @param {String} dataHrefBase the URL of the base location of the data to be
 * downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 * @param {String} reportFileName the file name of the report to use as a data
 * source for the data visualization.
 */
function TopPagesRealtime({ dataHrefBase, reportFileName }) {
  const dataURL = `${dataHrefBase}/${reportFileName}`;
  const ref = useRef(null);

  useEffect(() => {
    const initRealtimeChart = async () => {
      const result = await d3
        .select(ref.current)
        .datum({
          source: dataURL,
          block: ref.current,
        })
        .call(
          renderBlock
            .loadAndRender()
            .transform((d) => d.data)
            .on("render", (selection) => {
              selection
                .selectAll(".label")
                .each(function (d) {
                  d.text = this.innerText;
                })
                .html("")
                .text((d) => titleExceptions[d.page] || d.page_title);
            })
            .render(
              barChart()
                .label((d) => d.page_title)
                .value((d) => +d.active_visitors)
                .scale((values) =>
                  d3.scale
                    .linear()
                    .domain([0, 1, d3.max(values)])
                    .rangeRound([0, 1, 100]),
                )
                .format(formatters.addCommas),
            ),
        );
      return result;
    };
    initRealtimeChart().catch(console.error);
  }, []);

  return (
    <figure
      className="top-pages__realtime"
      data-source={dataURL}
      data-refresh="15"
      ref={ref}
    >
      <div className="data bar-chart"></div>
    </figure>
  );
}

TopPagesRealtime.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
  reportFileName: PropTypes.string.isRequired,
};

export default TopPagesRealtime;
