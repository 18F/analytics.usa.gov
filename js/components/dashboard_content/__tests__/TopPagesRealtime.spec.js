import React from "react";
import { render, screen, waitFor } from "@testing-library/react";

import DataLoader from "../../../lib/data_loader";
import TopPagesRealtime from "../TopPagesRealtime";

jest.mock("../../../lib/data_loader", () => ({
  ...jest.requireActual("../../../lib/data_loader"),
  loadJSON: jest.fn(),
}));

describe("TopPagesRealtime", () => {
  let component;
  let data;

  describe("when data is not loaded", () => {
    beforeEach(() => {
      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.resolve(null);
      });
      component = render(
        <TopPagesRealtime
          dataHrefBase="http://www.example.com/data/"
          reportFileName="foobar.json"
          numberOfListingsToDisplay={10}
          refreshSeconds={30}
        />,
      );
    });

    it("renders a component in loading state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data is loaded", () => {
    beforeEach(async () => {
      data = {
        name: "top-pages-realtime",
        query: {},
        meta: {
          name: "Top Pages (Live)",
          description:
            "The top 20 pages, measured by active onsite users, for all sites.",
        },
        data: [
          {
            page_title: "National Institute of Standards and Technology | NIST",
            active_visitors: "7689",
          },
          {
            page_title: "USPS.com® - USPS Tracking® Results",
            active_visitors: "7657",
          },
          {
            page_title: "Federal Student Aid",
            active_visitors: "3576",
          },
          {
            page_title: "USAJOBS - Search",
            active_visitors: "2242",
          },
          {
            page_title: "USPS.com® - Sign In",
            active_visitors: "1452",
          },
          {
            page_title:
              "Search Public Sex Offender Registries | Dru Sjodin National Sex Offender Public Website",
            active_visitors: "1420",
          },
          {
            page_title: "Informed Delivery",
            active_visitors: "1347",
          },
          {
            page_title: "NWS Radar",
            active_visitors: "1266",
          },
          {
            page_title: "USAJOBS - Job Announcement",
            active_visitors: "1226",
          },
          {
            page_title:
              "Official ESTA Application Website, U.S. Customs and Border Protection",
            active_visitors: "928",
          },
          {
            page_title: "Social Security",
            active_visitors: "714",
          },
          {
            page_title: "Welcome | USPS",
            active_visitors: "629",
          },
          {
            page_title: "SAM.gov",
            active_visitors: "566",
          },
          {
            page_title: "SAM.gov | Search",
            active_visitors: "564",
          },
          {
            page_title: "Log In | Federal Student Aid",
            active_visitors: "500",
          },
          {
            page_title: "National Weather Service",
            active_visitors: "485",
          },
          {
            page_title: "USPS.com® - USPS Tracking®",
            active_visitors: "440",
          },
          {
            page_title: "FAFSA® Application | Federal Student Aid",
            active_visitors: "429",
          },
          {
            page_title: "Home - My HealtheVet - My HealtheVet",
            active_visitors: "400",
          },
          {
            page_title:
              "Home | Dru Sjodin National Sex Offender Public Website",
            active_visitors: "400",
          },
        ],
        totals: {},
        taken_at: "2024-01-05T16:05:47.043Z",
      };

      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.resolve(data);
      });
      component = render(
        <TopPagesRealtime
          dataHrefBase="http://www.example.com/data/"
          reportFileName="foobar.json"
          numberOfListingsToDisplay={10}
          refreshSeconds={30}
        />,
      );
      await waitFor(() => screen.getByText("USAJOBS - Search"));
    });

    it("renders a component with data loaded", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data loading has an error", () => {
    const error = "you broke it";

    beforeEach(() => {
      console.error = jest.fn();
      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.reject(error);
      });
      component = render(
        <TopPagesRealtime
          dataHrefBase="http://www.example.com/data/"
          reportFileName="foobar.json"
          numberOfListingsToDisplay={10}
          refreshSeconds={30}
        />,
      );
    });

    it("renders a component in error state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });

    it("logs the error to console", () => {
      expect(console.error).toHaveBeenCalledWith(error);
    });
  });
});
