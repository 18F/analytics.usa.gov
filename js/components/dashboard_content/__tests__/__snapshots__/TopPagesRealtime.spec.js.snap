// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopPagesRealtime when data is loaded renders a component with data loaded 1`] = `
<DocumentFragment>
  <figure
    class="top-pages__realtime-bar-chart"
  >
    <div
      class="data bar-chart"
    >
      <div
        class="bin"
      >
        <div
          class="label"
        >
          National Institute of Standards and Technology | NIST
        </div>
        <div
          class="value"
        >
          7,689
        </div>
        <div
          class="bar"
          style="width: 0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          USPS.com® - USPS Tracking® Results
        </div>
        <div
          class="value"
        >
          7,657
        </div>
        <div
          class="bar"
          style="width: 0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          Federal Student Aid
        </div>
        <div
          class="value"
        >
          3,576
        </div>
        <div
          class="bar"
          style="width: 0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          USAJOBS - Search
        </div>
        <div
          class="value"
        >
          2,242
        </div>
        <div
          class="bar"
          style="width: 0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          USPS.com® - Sign In
        </div>
        <div
          class="value"
        >
          1,452
        </div>
        <div
          class="bar"
          style="width: 0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          Search Public Sex Offender Registries | Dru Sjodin National Sex Offender Public Website
        </div>
        <div
          class="value"
        >
          1,420
        </div>
        <div
          class="bar"
          style="width: 0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          Informed Delivery
        </div>
        <div
          class="value"
        >
          1,347
        </div>
        <div
          class="bar"
          style="width: 0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          NWS Radar
        </div>
        <div
          class="value"
        >
          1,266
        </div>
        <div
          class="bar"
          style="width: 0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          USAJOBS - Job Announcement
        </div>
        <div
          class="value"
        >
          1,226
        </div>
        <div
          class="bar"
          style="width: 0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          Official ESTA Application Website, U.S. Customs and Border Protection
        </div>
        <div
          class="value"
        >
          928
        </div>
        <div
          class="bar"
          style="width: 0%;"
        />
      </div>
    </div>
  </figure>
</DocumentFragment>
`;

exports[`TopPagesRealtime when data is not loaded renders a component in loading state 1`] = `
<DocumentFragment>
  <figure
    class="top-pages__realtime-bar-chart"
  >
    <div
      class="data bar-chart"
    />
  </figure>
</DocumentFragment>
`;

exports[`TopPagesRealtime when data loading has an error renders a component in error state 1`] = `
<DocumentFragment>
  <figure
    class="top-pages__realtime-bar-chart"
  >
    <div
      class="data bar-chart"
    />
  </figure>
</DocumentFragment>
`;
